### PR TITLE
feat: Record Column value Length at Segment level if the length is constant for all the Records

### PR DESCRIPTION
**Pull request by Macbeth98**
# Description
- Adds a new Field called `cValSize` that records the length of a constant Column Value for records at segment level.
- If the length of the column value is constant for all the records in the segment, then that value will be record, otherwise `-1` is set.
- `getCurrentRecordLength` uses this value if `+ve` otherwise, it proceeds through normal process.

# Testing
- Added a test case that tests the sizes.

# Checklist:

- [x] I have self-reviewed this PR.
- [x] I have removed all print-debugging and commented-out code that should not be merged.
- [x] I have added sufficient comments in my code, particularly in hard-to-understand areas.
- [x] I have formatted the code, if applicable. For Go, I have run `goimports -w .`.

### CODE REVIEW

File: pkg/segment/structs/segmeta.go
Comment (nkunal) [inline pkg/segment/structs/segmeta.go:29]: Since this is just meta and will get store once in segmeta per segment, then we can just go with a type of `int` , that should handle the cases where the string length is greater than 127 and we don't have to worry about overflows.
Plus it is not of big hit of 1 byte v/s 4 bytes for this var type

File: pkg/segment/writer/packer.go
Comment (nkunal) [inline pkg/segment/writer/packer.go:561]: remove commented code

File: pkg/segment/writer/packer.go
Comment (nkunal) [inline pkg/segment/writer/packer.go:687]: update the comment to something of ...
// column appearing first time in the middle of the wip, so past recNums will be filled with BackFill_CVAL_TYpe, so mark this as inconsistent

File: pkg/segment/writer/packer.go
Comment (nkunal) [inline pkg/segment/writer/packer.go:691]: we can just set `ss.AllSeenColumns[colName] = inSize` and return from here

File: pkg/segment/writer/packer.go
Comment (nkunal) [inline pkg/segment/writer/packer.go:689]: we don't need to set currentSize and just return from here. (note see below comment, we can just have 1 return statement for this whole if !ok block

File: pkg/segment/writer/packer.go
Comment (nkunal) [inline pkg/segment/writer/packer.go:708]: we won't need this if check based on above comments

File: pkg/segment/writer/packer.go
Comment (nkunal) [inline pkg/segment/writer/packer.go:710]: would be great if we can avoid type conversions, since this is the fast path code, understand the need of -1, but we can pick some other value for noting down inconsistency

File: pkg/segment/writer/segstore.go
Comment (nkunal) [inline pkg/segment/writer/segstore.go:74]: since we are repurposing this var for multiple things, maybe we should rename this to something like `AllSeenColumnsAndSizes` (you could find a better name than my suggestion)

File: pkg/segment/structs/segmeta.go
Comment (nkunal) [inline pkg/segment/structs/segmeta.go:29]: maybe we can call this `ConsistentCvalSize` .. something that denotes not the size of the whole column values but the consistency of the columnVal size across the whole segment

**Review by nkunal — COMMENTED** [review]
overall looks fantastic, a few naming nits and data type suggestion.
Great Job!!

File: pkg/segment/query/metadata/segmentmicroindex.go
Comment (Macbeth98) [reply pkg/segment/query/metadata/segmentmicroindex.go:271]: yes, it should be `ColSizeInfo`. I thought I named it `ColSizeInfo`. Correcting it.

File: pkg/segment/writer/packer.go
Comment (Macbeth98) [reply pkg/segment/writer/packer.go:691]: Yes, I did this way so that I can check the size. But now since there is no limit. I can just assign the size here and return.

File: pkg/segment/writer/packer.go
Comment (Macbeth98) [reply pkg/segment/writer/packer.go:710]: Then I think I can make the type as `uint32` and for inconsistency the value will be `0`.  As the size by default comes in `uint32`.

File: pkg/segment/reader/segread/segreader.go
Comment (AndrewHess) [inline pkg/segment/reader/segread/segreader.go:265]: What is `+ve`?

File: pkg/segment/structs/segmeta.go
Comment (AndrewHess) [inline pkg/segment/structs/segmeta.go:29]: nit: It might be worth adding a comment here to describe what `ConsistentCvalSize` means

**Review by AndrewHess — COMMENTED** [review]
This looks good; just some minor comments

@nkunal I think this is about ready to merge, if you want to do some performance testing before it gets merged

File: pkg/segment/reader/segread/segreader.go
Comment (Macbeth98) [reply pkg/segment/reader/segread/segreader.go:265]: I mean if the value is `>0`

**Comment by Macbeth98** [comment]
> This looks good; just some minor comments
> 
> @nkunal I think this is about ready to merge, if you want to do some performance testing before it gets merged

Please wait, I am doing an optimization where I am fetching the Column sizes before at `QuerySegmentRequests` rather than fetching it at `InitSharedMultiColumnReaders`. I am not exactly sure how much performance this will improve, but we don't need an extra lock and unlock through this way.

**Comment by Macbeth98** [comment]
@nkunal @AndrewHess Can you review the PR again? I have changed the getting of the Column sizes to segment level instead of calling it per block.  I have ran the benchmarks also with this latest PR.

```
Develop
Total Run: 626s
getCurrentRecordLength: 30s
1st run: 1m42s
2nd run: 1m47s

This PR
Total Run: 591s
getCurrentRecordLength: 20s
1st run: 1m36s
2nd run: 1m43s
```

File: pkg/segment/structs/segmeta.go
Comment (AndrewHess) [inline pkg/segment/structs/segmeta.go:30]: Using 1MB seems quite odd to me. Why not use 0 or the max uint32 value to indicate this special case?

Alternatively, maybe it would make sense to use a `utils.Option[uint32]` here instead (although with that I think you'd have to write the JSON encode/decode functions, which may cause problems for non-literal types, but perhaps you could only handle certain types like uint32 for now)

This `ColSizeInfo` is JSON encoded and then written to segmeta.json, so whatever you do, we're going to have to support that for a long time

**Comment by nkunal** [comment]
Maybe use uint32 max

*Best Regards*
*-- Kunal Nawale ,**Founder/CEO, SigScalr*

File: pkg/segment/structs/segmeta.go
Comment (Macbeth98) [reply pkg/segment/structs/segmeta.go:30]: Before we decided to use `1MB` because as we will not allow the value size more than `65_535` bytes. So the size will never reach `1MB`. So Should I continue using the same or change it to max of `uint32`?

Comment (AndrewHess) [reply pkg/segment/structs/segmeta.go:30]: Let's use the max `uint32`

**Review by AndrewHess — APPROVED** [review]

**Pull request merged**
